src/test/java/com/aws/greengrass/certificatemanager/certificate/CISShadowMonitorTest.java

```diff
@@ -49,172 +77,507 @@
-        CISShadowMonitor cisShadowMonitor = new CISShadowMonitor(mockMqttClient, mockConnection, mockShadowClient,
-                mockExecutor, SHADOW_NAME, mockConnectivityInfoProvider);
-
-        when(mockShadowClient.SubscribeToShadowDeltaUpdatedEvents(any(), any(), any(), any()))
-                .thenReturn(CompletableFuture.completedFuture(0));
-        when(mockShadowClient.SubscribeToGetShadowAccepted(any(), any(), any(), any()))
-                .thenReturn(CompletableFuture.completedFuture(0));
+    void GIVEN_CISShadowMonitor_WHEN_cis_shadow_changed_THEN_new_cert_generated() throws Exception {
+        CountDownLatch shadowDeltaUpdated = whenShadowDeltaUpdated(1);
+
         cisShadowMonitor.startMonitor();
+        cisShadowMonitor.addToMonitor(certificateGenerator);
+
+        publishDesiredShadowState(Utils.immutableMap("newState", 1));
+
+        assertTrue(shadowDeltaUpdated.await(5L, TimeUnit.SECONDS));
+        verify(certificateGenerator, times(1)).generateCertificate(any(), any());
+    }
 
-        ArgumentCaptor<ShadowDeltaUpdatedSubscriptionRequest> updateSubscribeRequest = ArgumentCaptor.forClass(
-                ShadowDeltaUpdatedSubscriptionRequest.class);
-        ArgumentCaptor<GetShadowSubscriptionRequest> getSubscribeRequest = ArgumentCaptor.forClass(
-                GetShadowSubscriptionRequest.class);
-        ArgumentCaptor<GetShadowRequest> publishRequest = ArgumentCaptor.forClass(GetShadowRequest.class);
+    @Test
+    void GIVEN_CISShadowMonitor_WHEN_cis_shadow_changed_multiple_times_THEN_new_cert_generated_multiple_times() throws Exception {
+        int numShadowChanges = 3;
+
+        CountDownLatch shadowDeltaUpdated = whenShadowDeltaUpdated(numShadowChanges);
+
+        cisShadowMonitor.startMonitor();
+        cisShadowMonitor.addToMonitor(certificateGenerator);
 
-        verify(mockShadowClient, times(1)).SubscribeToShadowDeltaUpdatedEvents(
-                updateSubscribeRequest.capture(), eq(QualityOfService.AT_LEAST_ONCE), any(),
-                any());
-        assertThat(updateSubscribeRequest.getValue().thingName, is(SHADOW_NAME));
-        verify(mockShadowClient, times(1)).SubscribeToGetShadowAccepted(
-                getSubscribeRequest.capture(), eq(QualityOfService.AT_LEAST_ONCE), any(),
-                any());
-        assertThat(getSubscribeRequest.getValue().thingName, is(SHADOW_NAME));
+        IntStream.range(0, numShadowChanges).forEach(i -> {
+            publishDesiredShadowState(Utils.immutableMap("newState", i));
+        });
 
-        verify(mockShadowClient, times(1)).PublishGetShadow(publishRequest.capture(),
-                eq(QualityOfService.AT_LEAST_ONCE));
-        assertThat(publishRequest.getValue().thingName, is(SHADOW_NAME));
+        assertTrue(shadowDeltaUpdated.await(5L, TimeUnit.SECONDS));
+        verify(certificateGenerator, times(numShadowChanges)).generateCertificate(any(), any());
```

Comment:
This might be okay, but it seems fairly coupled to the implementation. Might be better to test state rather than behavior in this case

Reply:
I'm not sure I see the coupling---I believe this is testing the high-level "interface" of the monitor:  if we get 3 shadow deltas, we'd expect 3 rotations.

maybe we need a separate integration test?

Reply:
Not necessarily. If you get 3 shadow deltas, you just expect that `generateCertificate` is called with the latest connectivity info. Imagine if you add some queueing de-dup logic to the code. This test will become non-deterministic